### PR TITLE
gpu: change 'none' allocation policy

### PR DESCRIPTION
**Pull request by tkatila**
With shared-dev-num and multiple i915s in the resource request, try to find as many individual GPUs to expose to the container.

Previously, with multiple i915 resources, it was typical to get only one GPU device in the container.

Fixes: #1377

### CODE REVIEW

File: cmd/gpu_plugin/gpu_plugin.go [hunk 1]
```diff
@@ -64,15 +64,58 @@ type cliOptions struct {
 
 type preferredAllocationPolicyFunc func(*pluginapi.ContainerPreferredAllocationRequest) []string
 
-// nonePolicy is used for allocating GPU devices randomly.
+// nonePolicy is used for allocating GPU devices randombly, while trying
```
Comment (uniemimu) [inline cmd/gpu_plugin/gpu_plugin.go:67]:
```suggestion
// nonePolicy is used for allocating GPU devices randomly, while trying
```

File: cmd/gpu_plugin/gpu_plugin.go [hunk 2]
```diff
@@ -64,15 +64,58 @@ type cliOptions struct {
...
+	if neededCount > 0 {
+		for _, deviceID := range req.AvailableDeviceIDs {
+			if _, found := selected[deviceID]; !found {
+				selected[deviceID] = true
+				neededCount--
+			}
+
+			if neededCount == 0 {
+				break
+			}
```
Comment (eero-t) [inline cmd/gpu_plugin/gpu_plugin.go:104]:
```suggestion
			if _, found := selected[deviceID]; !found {
				selected[deviceID] = true
				neededCount--

				if neededCount == 0 {
					break
				}
			}
```

File: cmd/gpu_plugin/gpu_plugin.go [hunk 3]
```diff
@@ -64,15 +64,58 @@ type cliOptions struct {
...
+
+		if _, found := devices[device]; !found {
+			devices[device] = true
+			selected[deviceID] = true
+			neededCount--
+		}
+
+		if neededCount == 0 {
+			break
+		}
```
Comment (eero-t) [inline cmd/gpu_plugin/gpu_plugin.go:90]:
```suggestion

			if neededCount == 0 {
				break
			}
		}
```

File: cmd/gpu_plugin/gpu_plugin.go (hunk 2 above)
Comment (eero-t) [inline cmd/gpu_plugin/gpu_plugin.go:104]:
```suggestion

				if neededCount == 0 {
					break
				}
			}
```

File: cmd/gpu_plugin/gpu_plugin.go [hunk 4]
```diff
@@ -64,15 +64,58 @@ type cliOptions struct {
...
 	klog.V(2).Info("Select nonePolicy for GPU device allocation")
 
-	deviceIds := req.AvailableDeviceIDs[0:req.AllocationSize]
+	devices := make(map[string]bool)
+	selected := make(map[string]bool)
+	neededCount := req.AllocationSize
 
-	klog.V(2).Infof("Allocate deviceIds: %q", deviceIds)
+	// When shared-dev-num is greater than 1, try to find as
+	// many independent GPUs to satisfy the request.
```
Comment (eero-t) [inline cmd/gpu_plugin/gpu_plugin.go:77]:
```suggestion
	// When shared-dev-num is greater than 1, try to find as
	// many independent GPUs as possible, to satisfy the request.
```

File: cmd/gpu_plugin/gpu_plugin.go [hunk 5]
```diff
@@ -64,15 +64,58 @@ type cliOptions struct {
...
+			selected[deviceID] = true
+			neededCount--
+		}
+
+		if neededCount == 0 {
+			break
+		}
+	}
+
+	// If there aren't enough of independent GPUs, add any not taken deviceIDs.
```
Comment (eero-t) [inline cmd/gpu_plugin/gpu_plugin.go:93]:
```suggestion
	// If there were not enough independent GPUs, use remaining untaken deviceIDs.
```

File: cmd/gpu_plugin/gpu_plugin.go [hunk 6]
```diff
@@ -64,15 +64,58 @@ type cliOptions struct {
...
+				neededCount--
+			}
+
+			if neededCount == 0 {
+				break
+			}
+		}
+	}
+
+	// Convert selected map into an array
```
[the remark this answers is not in the document]
Comment (tkatila) [reply cmd/gpu_plugin/gpu_plugin.go:108]: If there are not enough i915 resources, k8s scheduler shouldn't allow Pod to be scheduled.

**Review by uniemimu — APPROVED** [review]

File: cmd/gpu_plugin/gpu_plugin.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (eero-t) [reply cmd/gpu_plugin/gpu_plugin.go:108]: But does not allocation size come from scheduler, and the array from plugin itself, and isn't communication between those asynchronous?

If that is the case, then e.g. one of the GPUs could drop off between scheduler decision and GPU plugin receiving the request, in which case array would be one member short of what scheduler expects...

[the remark this answers is not in the document]
Comment (tkatila) [reply cmd/gpu_plugin/gpu_plugin.go:108]: I'm not 100% sure about all the phases or interactions, but the list of devices comes from kubelet. Scheduler selects a node which has enough i915 resources (I doubt that scheduler knows the exact deviceIDs) and indicates to the cluster that a Pod was assigned to a particular node. Kubelet detects a new Pending node and eventually calls GetPreferredAllocation in the plugin with the available IDs. Kubelet should have an up -to-date list of deviceIDs, and the count of that list should be in sync with the i915 resource count.

It should be also noted that this is GetPreferredAllocation. It's not certain Kubelet will use the result.

If a GPU drops during the scheduling flow, GetPreferredAllocation result will be short. Kubelet will do then something, I don't know what. :) Anyhow, the previous _none_ profile also would have returned a short result.

[the remark this answers is not in the document]
Comment (eero-t) [reply cmd/gpu_plugin/gpu_plugin.go:108]: Ok, if the list and alloc size come from together scheduler, then it should be fine.   At worst kubelet will then tell container runtime to map a device that has gone away, and it probably will then error out.

Both items coming from scheduler sounds like a lot of communication overhead though. If there's e.g. 10 GPUs and 100x share count, that's 1000 device info array that's sent to back to node on every request.  Is that really true?

[the remark this answers is not in the document]
Comment (eero-t) [reply cmd/gpu_plugin/gpu_plugin.go:108]: As to earlier _none_ profile code not checking things, and working fine...  I do not think we have a test-case for a potential race condition where GPU goes away between scheduler decision and GPU plugin handling it.  That sounds pretty tricky to test reliably...

[the remark this answers is not in the document]
Comment (tkatila) [reply cmd/gpu_plugin/gpu_plugin.go:108]: > Both items coming from scheduler sounds like a lot of communication overhead though. If there's e.g. 10 GPUs and 100x share count, that's 1000 device info array that's sent to back to node on every request. Is that really true?

I'm not 100% sure. I believe scheduler (and its plugins/extenders) mostly work with numbers when doing their decisions. Kubelet (and its managers) then work with the actual device IDs.

**Comment by eero-t** [comment]
@uniemimu Can this change interfere with: https://github.com/intel/platform-aware-scheduling/blob/master/gpu-aware-scheduling/docs/usage.md#enforcing-same-gpu-to-multiple-containers-within-pod ?

**Comment by uniemimu** [comment]
> @uniemimu Can this change interfere with: https://github.com/intel/platform-aware-scheduling/blob/master/gpu-aware-scheduling/docs/usage.md#enforcing-same-gpu-to-multiple-containers-within-pod ?

No, this is totally separate from GAS.

File: cmd/gpu_plugin/gpu_plugin.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (eero-t) [reply cmd/gpu_plugin/gpu_plugin.go:108]: Ok, kubelet in same node being out of sync is much less likely.  And unlike the earlier implementation, this won't any more crash if they are not in sync, so this is clear improvement in that respect!

**Pull request merged**